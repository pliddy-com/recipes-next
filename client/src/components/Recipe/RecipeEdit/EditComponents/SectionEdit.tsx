import { useState } from 'react';

import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';

import ListEdit from './ListEdit';
import TextEdit from './TextEdit';

import { toSlug } from 'lib/utils';

import { IRecipeSection } from 'types/content';

interface ISectionEdit {
  id: string;
  label: string;
  onChange({ id, value }: { id: string; value: IRecipeSection[] }): void;
  sectionList: IRecipeSection[];
}

const SectionEdit = ({ id, label, onChange, sectionList }: ISectionEdit) => {
  const [sectionData, setSectionData] = useState<IRecipeSection[]>(sectionList);

  const updateSection = ({
    index,
    value
  }: {
    index: number;
    value: (string | null)[];
  }) => {
    const sections = [...sectionData];
    const section = { ...sections[index] };
    section.sectionItems = value;
    sections[index] = section;

    setSectionData(sections);
    onChange({ id, value: sections });
  };

  const updateTitle = ({ index, value }: { index: number; value: string }) => {
    const sections = [...sectionData];
    const section = { ...sections[index] };
    section.sectionTitle = value;
    sections[index] = section;

    setSectionData(sections);
    onChange({ id, value: sections });
  };

  return (
    <>
      <Typography variant="h2">{label}</Typography>
      {sectionData.map((section, index) => {
        const { sectionItems, sectionTitle } = section;
        return sectionItems && sectionTitle ? (
          <Box key={`section-title-${index}`}>
            <TextEdit
              className="bold"
              id={`${toSlug(sectionTitle)}-title`}
              label={sectionTitle}
              onChange={(e) => updateTitle({ index, value: e.target.value })}
              value={sectionTitle}
            />

            <ListEdit
              heading="h3"
              id={toSlug(sectionTitle).toLowerCase()}
              label={sectionTitle}
              list={sectionItems}
              onChange={({ value }) => updateSection({ index, value })}
            />
          </Box>
        ) : null;
      })}
    </>
  );
};

export default SectionEdit;
