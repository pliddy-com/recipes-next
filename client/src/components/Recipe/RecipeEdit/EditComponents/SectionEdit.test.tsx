// import testing-library methods
import { fireEvent, render } from '@testing-library/react';

import SectionEdit from './SectionEdit';

jest.mock('./TextEdit');

describe('SectionEdit', () => {
  describe('when there is content', () => {
    it('renders the component', async () => {
      const props = {
        id: 'Test Section',
        label: 'Section Label',
        onChange: jest.fn(),
        sectionList: [
          {
            sectionTitle: 'Section Title',
            sectionItems: ['item 1', 'item2']
          }
        ]
      };

      const testTitle = 'Test Title';
      const testItem = 'item 3';

      const { asFragment, queryByLabelText } = render(
        <SectionEdit {...props} />
      );

      expect(asFragment()).toMatchSnapshot();

      // assert that onChange executes
      const titleInput = queryByLabelText('Section Title');
      titleInput &&
        fireEvent.change(titleInput, {
          target: { value: testTitle }
        });

      const firstInput = queryByLabelText('Section Title 1');
      firstInput &&
        fireEvent.change(firstInput, {
          target: { value: testItem }
        });
    });
  });

  describe('when there are missing section props', () => {
    it('does not render the component', () => {
      const props = {
        id: 'Test Section',
        label: 'Section Label',
        onChange: jest.fn(),
        sectionList: [
          {
            sectionTitle: null,
            sectionItems: ['item 1', 'item2']
          }
        ]
      };
      const { asFragment } = render(<SectionEdit {...props} />);

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
