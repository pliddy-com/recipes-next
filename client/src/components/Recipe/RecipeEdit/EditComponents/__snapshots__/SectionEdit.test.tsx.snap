// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionEdit when there are missing section props does not render the component 1`] = `
<DocumentFragment>
  <h2
    class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
  >
    Section Label
  </h2>
</DocumentFragment>
`;

exports[`SectionEdit when there is content renders the component 1`] = `
<DocumentFragment>
  <h2
    class="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
  >
    Section Label
  </h2>
  <div
    class="MuiBox-root css-0"
  >
    <div
      data-testid="mock-text-field-Section Title"
    >
      <label>
        Section Title
      </label>
      <input
        value="Section Title"
      />
    </div>
    <div
      class="listEdit css-nen11g-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="section-title-1"
          id="section-title-1-label"
        >
          Section Title 1
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
            id="section-title-1"
            type="text"
            value="item 1"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="move item up"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleUpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0 2c5.52 0 10-4.48 10-10S17.52 2 12 2 2 6.48 2 12s4.48 10 10 10zm-1-10v4h2v-4h3l-4-4-4 4h3z"
                />
              </svg>
            </button>
            <button
              aria-label="move item down"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 10V8h-2v4H8l4 4 4-4h-3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="remove item"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CancelIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Section Title 1
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root field css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for="section-title-2"
          id="section-title-2-label"
        >
          Section Title 2
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
            id="section-title-2"
            type="text"
            value="item2"
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
          >
            <button
              aria-label="move item up"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleUpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0 2c5.52 0 10-4.48 10-10S17.52 2 12 2 2 6.48 2 12s4.48 10 10 10zm-1-10v4h2v-4h3l-4-4-4 4h3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="move item down"
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowCircleDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 10V8h-2v4H8l4 4 4-4h-3z"
                />
              </svg>
            </button>
            <button
              aria-label="remove item"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="CancelIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Section Title 2
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        aria-label="add item"
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddCircleIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
            />
          </svg>
        </span>
        Add Item
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
